File: frontend/src/lib/commands.ts
import { get, send, asList, pick } from "./http";

export interface SlashCommand {
  name: string;
  description: string;
  category: string;
  usage?: string;
}

function toCmd(c: Record<string, unknown>): SlashCommand {
  return {
    name: String(pick(c, ["name", "command"], "")),
    description: String(pick(c, ["description", "summary"], "")),
    category: String(pick(c, ["category"], "general")),
    usage: typeof c.usage === "string" ? c.usage : undefined,
  };
}

export async function listCommands(): Promise<SlashCommand[]> {
  const d = await get<unknown>("/commands");
  return asList(d, ["commands", "data"]).map((c) => toCmd(c));
}

export async function searchCommands(q: string): Promise<SlashCommand[]> {
  const d = await get<unknown>(`/commands/search?q=${encodeURIComponent(q)}`);
  return asList(d, ["commands", "results", "data"]).map((c) => toCmd(c));
}

export async function commandCategories(): Promise<Array<{ name: string; count: number }>> {
  try {
    const d = await get<unknown>("/commands/categories");
    const list = asList(d, ["categories", "data"]);
    if (list.length > 0) {
      return list.map((c) => ({
        name: String(pick(c, ["name", "category"], "")),
        count: Number(pick(c, ["count", "total"], 0)),
      }));
    }
    const rec = (d as Record<string, unknown>).categories;
    if (rec && typeof rec === "object") {
      return Object.entries(rec as Record<string, unknown>).map(([name, count]) => ({
        name,
        count: typeof count === "number" ? count : 0,
      }));
    }
  } catch {
    /* ignore */
  }
  return [];
}

/** Run a slash command (e.g. "/goal status"). Returns a human-readable summary. */
export async function executeCommand(command: string, context?: Record<string, unknown>): Promise<string> {
  const d = await send<Record<string, unknown>>("/commands/execute", "POST", {
    command,
    ...(context ? { context } : {}),
  });
  const msg = pick<string | null>(d, ["message", "result", "output", "summary"], null);
  if (msg) return String(msg).slice(0, 4000);
  return JSON.stringify(d, null, 2).slice(0, 4000);
}

File: backend/app/gateway/routers/commands.py
"""Gateway API Router for Master Slash Commands.

Exposes endpoints for:
- Querying the global Master Slash Command registry (418+ commands across 28 families)
- Category listing and counts
- Substring and capability search across commands and descriptions
- Direct command dispatch and intent resolution for the UI/Agents
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Query, Request
from pydantic import BaseModel, Field

from deerflow.commands import CommandCategory, command_registry

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/commands", tags=["commands"])


class CommandExecuteRequest(BaseModel):
    command: str = Field(..., min_length=1, description="Slash command line to execute, e.g. '/goal status' or '/plan'")
    context: Optional[Dict[str, Any]] = Field(default=None, description="Optional execution context such as thread_id, agent_id, or options")


@router.get("")
async def list_commands(
    category: Optional[str] = Query(default=None, description="Filter by command category (e.g. 'core', 'mission', 'swarm')"),
    core_only: bool = Query(default=False, description="Filter only core commands"),
) -> Dict[str, Any]:
    """Lists all registered slash commands with optional category or core filter."""
    cat_enum: Optional[CommandCategory] = None
    if category:
        try:
            cat_enum = CommandCategory(category.lower())
        except ValueError:
            cat_enum = None

    commands = await asyncio.to_thread(command_registry.list_commands, category=cat_enum, only_core=core_only)
    return {
        "total": len(commands),
        "category": category,
        "core_only": core_only,
        "commands": [c.to_dict() for c in commands],
    }


@router.get("/categories")
async def get_categories() -> Dict[str, Any]:
    """Returns all 28 command categories and their command counts."""
    categories = await asyncio.to_thread(command_registry.get_categories)
    return {
        "categories": categories,
        "total_categories": len(categories),
    }


@router.get("/search")
async def search_commands(
    q: str = Query(..., min_length=1, description="Search query string"),
) -> Dict[str, Any]:
    """Searches commands by name, description, or category."""
    results = await asyncio.to_thread(command_registry.search, query=q)
    return {
        "query": q,
        "total": len(results),
        "commands": [c.to_dict() for c in results],
    }


@router.post("/execute")
async def execute_command(payload: CommandExecuteRequest) -> Dict[str, Any]:
    """Dispatches a slash command intent to the runtime."""
    result = await asyncio.to_thread(
        command_registry.execute,
        command_line=payload.command,
        context=payload.context,
    )
    return result.to_dict()
